asim: impl rebalance interface

kvoli opened the pull request:
This patch adds range rebalancer simulation. `RelocateRange` and
`AdminTransferLease` methods are implemented within the allocator
simulator.

These methods are non-blocking. `AdminTransferLease` applies with some
delay and `RelocateRange` is applied over multiple ticks, iteratively
calling `RelocateOne` to determine the next set of replication changes
to apply.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85433)
<!-- Reviewable:end -->

kvoli committed:
asim: move queues into separate pkg

This patch separates out the replicate queue and split queue into
separate files. This patch also moves the respective queues and the base
queue into a separate package within asim.

Release note: None

kvoli committed:
asim: simulate relocate range

This patch adds range rebalancer simulation. `RelocateRange` and
`AdminTransferLease` methods are implemented within the allocator
simulator.

These methods are non-blocking. `AdminTransferLease` applies with some
delay and `RelocateRange` is applied over multiple ticks, iteratively
calling `RelocateOne` to determine the next set of replication changes
to apply.

Release note: None

kvoli commented:
merged into #86249
